include/oneapi/internal/execution_context_sycl.h

```diff
@@ -84,12 +150,17 @@ class OpenClKernelFactory : public Base, public ClKernelFactoryIface
 
         services::String keyCache = _clProgramRef->getName();
         keyCache.add(kernelName);
-        const uint64_t id = hash(keyCache.c_str()) % SIZE_CACHE_KERNEL;
+        uint64_t id = hash(keyCache.c_str()) % SIZE_CACHE_KERNEL;
         // TODO: Thread safe?
 
-        if (_kernelCache[id])
+        while (_kernelCache[id].getKernel() && strcmp(_kernelCache[id].getName(), kernelName) != 0)
```

Comment:
It duplicates the code above. Again, it can be solved by introducing new entity `KernelCache`/`ProgramCache` or smth like this.